File: src/main/java/me/andre111/d20server/model/ServerEntityManager.java
package me.andre111.d20server.model;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.function.BiConsumer;
import java.util.function.Consumer;
import java.util.function.Predicate;

import com.google.gson.reflect.TypeToken;

import me.andre111.d20common.message.game.entity.AddEntity;
import me.andre111.d20common.message.game.entity.ClearEntities;
import me.andre111.d20common.message.game.entity.RemoveEntity;
import me.andre111.d20common.message.game.entity.UpdateEntityProperties;
import me.andre111.d20common.model.Entity;
import me.andre111.d20common.model.EntityManager;
import me.andre111.d20common.model.Property;
import me.andre111.d20common.model.def.EntityDefinition;
import me.andre111.d20common.model.profile.Profile;
import me.andre111.d20common.model.property.Access;
import me.andre111.d20common.util.Utils;
import me.andre111.d20server.service.MessageService;
import me.andre111.d20server.service.SaveService;
import me.andre111.d20server.service.UserService;

public class ServerEntityManager implements EntityManager {
	protected final String type;
	protected final boolean synched;
	protected final Access addRemoveAccess;
	
	private final Map<Long, Entity> entities;
	
	private final List<Consumer<Map<Long, Entity>>> listeners = new ArrayList<>();
	private final List<Consumer<Entity>> entityListeners = new ArrayList<>();
	private final List<BiConsumer<Long, Entity>> removalListeners = new ArrayList<>();
	
	private boolean saveEnabled = true;
	
	public ServerEntityManager(String type, EntityDefinition def) {
		this.type = type;
		this.synched = def.settings().synched();
		this.addRemoveAccess = def.settings().addRemoveAccess();
		
		Map<Long, Entity> loadedEntities = Utils.readJson("entity."+type, TypeToken.getParameterized(HashMap.class, Long.class, Entity.class).getType());
		entities = loadedEntities != null ? loadedEntities : new HashMap<>();
	}
	
	@Override
	public final Entity find(long id) {
		return entities.get(id);
	}
	
	@Override
	public final boolean has(long id) {
		return entities.containsKey(id);
	}
	
	@Override
	public final List<Entity> all() {
		return new ArrayList<>(entities.values());
	}
	
	@Override
	public final Map<Long, Entity> map() {
		return Collections.unmodifiableMap(entities);
	}

	@Override
	public final void add(Entity entity) {
		entities.put(entity.id(), entity);
		if(saveEnabled) Utils.saveJson("entity."+type, entities);
		
		UserService.forEach(profile -> {
			if(entity.canView(profile)) MessageService.send(new AddEntity(entity), profile);
		});
		
		notifyListeners();
		for(var listener : entityListeners) {
			listener.accept(entity);
		}
	}

	@Override
	public final void remove(long id) {
		if(!entities.containsKey(id)) return;
		
		Entity entity = entities.remove(id);
		if(saveEnabled) Utils.saveJson("entity."+type, entities);
		
		UserService.forEach(profile -> {
			MessageService.send(new RemoveEntity(type, id), profile);
		});
		
		notifyListeners();
		for(var listener : removalListeners) {
			listener.accept(id, entity);
		}
	}

	@Override
	public final void updateProperties(long id, Map<String, Property> map, Access accessLevel) {
		Entity entity = find(id);
		if(entity == null) return;
		
		// keep track of all profiles that could see this entity
		Set<Profile> couldView = new HashSet<>();
		for(Profile profile : UserService.getAllConnectedProfiles()) {
			if(entity.canView(profile)) couldView.add(profile);
		}
		
		// transfer properties respecting access settings and keeping track of which changed
		Map<String, Property> changedProperties = new HashMap<>();
		for(Map.Entry<String, Property> e : map.entrySet()) {
			Property ownProperty = entity.prop(e.getKey());
			if(ownProperty == null) continue; // server discards new/unknown properties from client
			if(!e.getValue().hasValidValue()) continue; // server discards properties with invalid value
			
			// transfer value
			if(accessLevel.ordinal() >= ownProperty.getEditAccess().ordinal()) {
				try {
					e.getValue().transferTo(ownProperty);
					changedProperties.put(e.getKey(), ownProperty);
				} catch(UnsupportedOperationException ex) {
					ex.printStackTrace(); //TODO: how to handle incorrect property updates
				}
			}
			
			// transfer access (GM only, and do not allow changing of SYSTEM level access)
			if(accessLevel == Access.GM) {
				if(e.getValue().getViewAccess() != Access.SYSTEM && ownProperty.getViewAccess() != Access.SYSTEM && e.getValue().getViewAccess() != ownProperty.getViewAccess()) {
					ownProperty.setViewAccess(e.getValue().getViewAccess());
					changedProperties.put(e.getKey(), ownProperty);
				}
				if(e.getValue().getEditAccess() != Access.SYSTEM && ownProperty.getEditAccess() != Access.SYSTEM  && e.getValue().getEditAccess() != ownProperty.getEditAccess()) {
					ownProperty.setEditAccess(e.getValue().getEditAccess());
					changedProperties.put(e.getKey(), ownProperty);
				}
			}
		}

		// save and transfer (depending on (changing) access: add, remove or only changed properties)
		if(saveEnabled) SaveService.requestSave(type); // request (async) save, as that could take some time
		UserService.forEach(profile -> {
			if(entity.canView(profile) && !couldView.contains(profile)) {
				MessageService.send(new AddEntity(entity), profile);
			} else if(!entity.canView(profile) && couldView.contains(profile)) {
				MessageService.send(new RemoveEntity(type, id), profile); 
			} else if(entity.canView(profile)) {
				MessageService.send(new UpdateEntityProperties(type, id, changedProperties), profile);
			}
		});
		
		notifyListeners();
		for(var listener : entityListeners) {
			listener.accept(entity);
		}
	}

	
	public final void addListener(Consumer<Map<Long, Entity>> listener) {
		listeners.add(listener);
	}
	
	public final void addEntityListener(Consumer<Entity> listener) {
		entityListeners.add(listener);
	}
	
	public final void addRemovalListener(BiConsumer<Long, Entity> listener) {
		removalListeners.add(listener);
	}
	
	private final void notifyListeners() {
		Map<Long, Entity> unmodifiable = Collections.unmodifiableMap(entities);
		for(var listener : listeners) {
			listener.accept(unmodifiable);
		}
	}
	
	
	public final boolean canAddRemove(Profile profile, Entity entity) {
		Access accessLevel = entity.getAccessLevel(profile);
		return addRemoveAccess.ordinal() <= accessLevel.ordinal();
	}
	
	protected final int getAccessibleCount(Profile profile) {
		int count = 0;
		for(Entity entity : entities.values()) {
			if(entity.canView(profile)) {
				count++;
			}
		}
		return count;
	}
	
	protected final void fullSync(Profile profile) {
		MessageService.send(new ClearEntities(type), profile);
		for(Entity entity : entities.values()) {
			if(entity.canView(profile)) {
				MessageService.send(new AddEntity(entity), profile);
			}
		}
	}
	
	protected final void removeAll(Predicate<Entity> predicate) {
		for(Entity e : new ArrayList<>(entities.values())) {
			if(predicate.test(e)) remove(e.id());
		}
	}
	
	public final void setSaveEnabled(boolean saveEnabled) {
		this.saveEnabled = saveEnabled;
		if(saveEnabled) {
			SaveService.requestSave(type);
		}
	}
	
	public final void performSave() {
		Utils.saveJson("entity."+type, entities); // request (async) save, as that could take some time
	}
}

File: core/server/entity/server-entity-manager.js
// @ts-check
import { EntityManager } from '../../common/entity/entity-managers.js';
import { Entity } from '../../common/common.js';
import { Access, Role } from '../../common/constants.js';
import { AddEntities, ClearEntities, RemoveEntity, UpdateEntityProperties } from '../../common/messages.js';
import { MessageService } from '../service/message-service.js';
import { UserService } from '../service/user-service.js';

import { chunk, fromJson, toJson } from '../../common/util/datautil.js';
import nedb from '@rmanibus/nedb'; // NOTE: This uses a special 'updated' nedb fork (with updated dependencies)
import fs from 'fs-extra';
import { PARAMETERS } from '../parameters.js';
import { Events } from '../../common/events.js';

export class ServerEntityManager extends EntityManager {
    synched;
    addRemoveAccess;

    entities;
    db;

    saveEnabled = true;
    saveOperations = 0;

    constructor(name, type, entityDefinition, cb) {
        super(name, type);

        this.synched = entityDefinition.settings.synched;
        this.addRemoveAccess = entityDefinition.settings.addRemoveAccess;

        // load entities (into map)
        this.entities = {};
        this.db = new nedb({ filename: './' + PARAMETERS.datadir + '/entity/' + name + '.db', autoload: true });
        this.db.persistence.setAutocompactionInterval(1000 * 60 * 60); // auto compact every hour to avoid files getting to large during long sessions
        this.db.find({}, (err, docs) => {
            for (const doc of docs) {
                this.entities[doc._id] = fromJson(doc.json);
                this.entities[doc._id].manager = this.getName();
                this.entities[doc._id].onAdd();
            }
            if (cb) cb();
        });
    }

    find(id) {
        return this.entities[String(id)];
    }

    has(id) {
        const entity = this.entities[String(id)]
        return entity != null && entity != undefined;
    }

    all() {
        return Object.values(this.entities); //TODO: might need the readOnlyAll system from old client
    }

    map() {
        return this.entities; //TODO: should be an unmodifiable view
    }

    add(entity) {
        if (!(entity instanceof Entity)) throw new Error('Can only add instances of entity');
        if (entity.getType() != this.getType()) throw new Error('Entity is of wrong type');

        entity.manager = this.getName();
        entity.onAdd();

        const id = String(entity.getID());
        this.entities[id] = entity;
        if (this.saveEnabled) {
            this.saveOperations++;
            const doc = {
                _id: id,
                json: toJson(entity, false)
            };
            this.db.update({ _id: id }, doc, { upsert: true }, err => {
                this.saveOperations--;
                //TODO: error handler
            });
        }

        UserService.forEach(profile => {
            if (entity.canView(profile)) this.syncEntity(profile, entity);
        });

        this.triggerEvent('added', entity);
    }

    remove(id) {
        const entity = this.find(id);
        if (!entity) return;

        entity.onRemove();

        delete this.entities[String(id)];
        if (this.saveEnabled) {
            this.saveOperations++;
            this.db.remove({ _id: String(id) }, {}, err => {
                this.saveOperations--;
                //TODO: error handler
            });
        }

        UserService.forEach(profile => {
            MessageService.send(new RemoveEntity(this.getName(), id), profile);
        });

        this.triggerEvent('removed', entity);
    }

    updateProperties(id, map, accessLevel) {
        const entity = this.find(id);
        if (!entity) return;

        // keep track of all profiles that could see this entity
        var couldView = [];
        UserService.forEach(profile => {
            if (entity.canView(profile)) couldView.push(profile);
        });

        // transfer properties respecting access settings and keeping track of which changed
        // check properties first for any obviously invalid changes
        var propertiesToChange = {};
        for (const [key, value] of Object.entries(map)) {
            if (!entity.has(key)) continue; // server discards new/unknown properties from client
            if (!entity.isValidValue(key, value)) continue; // server discards properties with invalid value

            // transfer value
            if (entity.canEditProperty(key, accessLevel)) {
                propertiesToChange[key] = value;
            }
        }

        // call event
        const event = Events.trigger('modify_' + this.getType(), {
            entity: entity,
            propertiesToChange: propertiesToChange,
            accessLevel: accessLevel,
            manager: this
        }, true);
        if (event.canceled) return;

        // apply changes (checking again for invalid changes made by event callbacks)
        var changedProperties = {};
        for (const [key, value] of Object.entries(propertiesToChange)) {
            if (!entity.has(key)) continue; // server discards new/unknown properties from client
            if (!entity.isValidValue(key, value)) continue; // server discards properties with invalid value

            // transfer value
            if (entity.canEditProperty(key, accessLevel)) {
                try {
                    entity.setInternal(key, value);
                    changedProperties[key] = value;

                    const updated = entity.onPropertyChange(key);
                    for (const [ukey, uvalue] of Object.entries(updated)) {
                        changedProperties[ukey] = uvalue;
                    }
                } catch (error) {
                    //TODO: how to handle incorrect property updates
                    console.log(`Recieved invalid property value: ${error}`);
                }
            }
        }

        // save
        if (this.saveEnabled) {
            this.saveOperations++;
            const doc = {
                _id: String(id),
                json: toJson(entity, false)
            };
            this.db.update({ _id: String(id) }, doc, { upsert: true }, err => {
                this.saveOperations--;
                //TODO: error handler
            });
        }

        // transfer (depending on (changing) access: add, remove or only changed properties)
        UserService.forEach(profile => {
            if (entity.canView(profile) && !couldView.includes(profile)) {
                this.syncEntity(profile, entity);
            } else if (!entity.canView(profile) && couldView.includes(profile)) {
                MessageService.send(new RemoveEntity(this.getName(), id), profile);
            } else if (entity.canView(profile)) {
                MessageService.send(new UpdateEntityProperties(this.getName(), id, changedProperties), profile);
            }
        });

        this.triggerEvent('modified', entity);
    }

    canView(profile) {
        if (profile.role == Role.SYSTEM) return true;

        // global always true, contained return parentEntity.canView(profile)
        if (this.parentEntity) {
            // special case: maps -> only sync content of current map
            if (this.parentEntity.getType() == 'map') {
                return this.parentEntity.getID() == profile.getCurrentMap();
            }

            return this.parentEntity.canView(profile);
        } else {
            return true;
        }
    }

    canAddRemove(profile, entity) {
        const accessLevel = entity.getAccessLevel(profile);
        return Access.matches(this.addRemoveAccess, accessLevel);
    }

    getAccessibleCount(profile) {
        if (!this.canView(profile)) return 0;

        var count = 0;
        for (const entity of Object.values(this.entities)) {
            if (entity.canView(profile)) count++;
        }
        return count;
    }

    fullSync(profile) {
        if (!this.canView(profile)) return;

        MessageService.send(new ClearEntities(this.getName(), this.getType()), profile);

        // collect entities
        var entitiesToSync = [];
        for (const entity of Object.values(this.entities)) {
            if (entity.canView(profile)) entitiesToSync.push(entity);
        }
        if (entitiesToSync.length == 0) return;

        // send them in ONE message
        // TODO: maybe splitting them up into batches of X would be better -> try to find a good balance
        //MessageService.send(new AddEntities(this.getName(), entitiesToSync), profile);
        for (const split of chunk(entitiesToSync, 20)) {
            MessageService.send(new AddEntities(this.getName(), split), profile);
        }

        // iterate contained managers and perform full sync
        for (const containedEntityType of entitiesToSync[0].getDefinition().settings.containedEntities) {
            for (const entity of entitiesToSync) {
                const manager = entity.getContainedEntityManager(containedEntityType);
                if (manager) manager.fullSync(profile);
            }
        }
    }

    syncEntity(profile, entity) {
        MessageService.send(new AddEntities(this.getName(), [entity]), profile);

        // iterate contained managers and perform full sync
        for (const containedEntityType of entity.getDefinition().settings.containedEntities) {
            const manager = entity.getContainedEntityManager(containedEntityType);
            if (manager) manager.fullSync(profile);
        }
    }

    removeAll(predicate) {
        for (const entity of Array.from(Object.values(this.entities))) {
            if (predicate(entity)) this.remove(entity.getID());
        }
    }

    setSaveEnabled(saveEnabled) {
        this.saveEnabled = saveEnabled;

        // perform full save on reenable
        if (saveEnabled) {
            for (const [id, entity] of Object.entries(this.entities)) {
                const doc = {
                    _id: id,
                    json: toJson(entity, false)
                };

                this.db.update({ _id: id }, doc, { upsert: true }, err => { }); //TODO: error handling
            }
            this.db.persistence.compactDatafile();
        }
    }

    onDelete() {
        // delete db file
        if (fs.existsSync('./data/entity/' + this.getName() + '.db')) {
            fs.remove('./data/entity/' + this.getName() + '.db');
        }
    }

    isSaving() {
        return this.saveOperations > 0 || !this.db.executor.queue.idle();
    }
}
